data/man/man1/wal.1
.TH WAL 1  pywal16
.SH NAME
wal \- Generate colorschemes on the fly
.SH SYNOPSIS
.B wal
.RB [ \-hclnqrRstvwe ]
.RB [ -a
.IR "alpha" ]
.RB [ -b
.IR background ]
.RB [ --fg
.IR foreground ]
.RB [ --backend
.IR [backend] ]
.RB [ --out-dir
.IR [out_dir] ]
.RB [ --theme
.IR /path/to/file
or
.IR theme_name ]
.RB [ --iterative ]
.RB [ --cols16
.IR [method] ]
.RB [ --recursive ]
.RB [ --saturate
.IR -1.0 - 1.0 ]
.RB [ --preview ]
.RB [ --vte ]
.RB [ -i
.IR "/path/to/img.jpg" ]
.RB [ -o
.IR "script_name" ]
.RB [ -p
.IR "theme_name" ]
.RB [ --contrast
.IR [1.0-21.0] ]
.SH DESCRIPTION
wal is a tool that generates a color palette from the dominant colors in an image. It then applies the colors system-wide and on-the-fly in all of your favourite programs.
.SH OPTIONS
.TP
.B "\-h, \-\-help"
Print brief usage information to standard output and exit.

.TP
.BI "\-a " alpha
Set terminal background transparency to
.IR alpha . 
The value must be an integer between 0 and 100.
 *Only works in URxvt and terminals that implement the OSC-11 interface*

.TP
.BI "\-b " background
Custom
.IR background
color to use.

.TP
.BI "\-\-fg " foreground
Custom
.IR foreground
color to use.

.TP
.BI "\-\-backend " backend
Which color backend to use. Use 'wal \-\-backend' to list backends.

.TP
.BI "\-\-out-dir " /path/to/output\ dir
Directory to output the built template files. 
By default wal will export the built templates to 
.IR $XDG_CACHE_HOME/wal/ 

The output dir for the templates can also be set with the
.IR $PYWAL_CACHE_DIR
env var, using this flag overrides the internal
.IR CACHE_DIR
variable taking precedence over $PYWAL_CACHE_DIR.

.TP
.BI "\-f, \-\-theme " /path/to/file\ or\ theme_name
Which colorscheme file to use. Use 'wal \-\-theme' to list builtin and user themes.

.TP
.BI "\-\-iterative "
When pywal is given a directory as input and this flag is used: Go through the images in order instead of shuffled.

.TP
.BI "\-\-cols16 " method
Use 16 color mode, this will generate 16 color themes or convert 9 color themes to 16 colors.
Method is one of
.IR darken
,
.IR lighten
or
.IR dual
default:
.IR darken

.TP
.B "\-\-recursive "
When pywal is given a directory as input and this flag is used: Search for images recursively in subdirectories instead of the root only.

.TP
.BI "\-\-saturate " [-1.0 - 1.0]
Set the color saturation.
Interpret as percentage, ie: 1.00 = 100%, 0.50 = 50%,

.TP
.B "\-\-preview "
Print the current color palette.

.TP
.B "\-\-vte "
Fix text-artifacts printed in VTE terminals.

.TP
.B "\-c "
Delete all cached colorschemes.

.TP
.BI "\-i " "/path/to/img.jpg"
Which image or directory to use.

.TP
.B "\-l "
Generate a light colorscheme.

.TP
.B "\-n "
Skip setting the wallpaper.

.TP
.BI "\-o " "script_name"
External script to run after "wal".

.TP
.BI "\-p " "theme_name"
permanently save theme to
.IR $XDG_CONFIG_HOME/wal/colorschemes
with the specified
.IR theme_name .

.TP
.B "\-q "
Quiet mode, don't print anything.

.TP
.B "\-r "
\'wal -r\' is deprecated: Use (cat ~/.cache/wal/sequences &) instead.

.TP
.B "\-R "
Restore previous colorscheme.

.TP
.B "\-s "
Skip changing colors in terminals.

.TP
.B "\-t "
Skip changing colors in tty.

.TP
.B "\-v "
Print "wal" version.

.TP
.B "\-w "
Use last used wallpaper for color generation.

.TP
.B "\-e "
Skip reloading xrdb/i3/sway/polybar

.TP
.IB "\-\-contrast " [1.0-21.0]
Specify a minimum contrast ratio between palette colors and the source image according to W3 contrast specifications. Values between 1.5-4.5 typically work best.

.SH CONFIGURATION
.P
Wal is configured through a handful of ENV VARS, the following ones are used:

.TP
.B "XDG_CONFIG_HOME"
Parent directory to the user wal/templates dir.

.TP
.B "PYWAL_CACHE_DIR"
Explicitly set the cache dir for the built templates, this has precedence over the XDG_CACHE_HOME/wal dir.

.TP
.B "NO_FUN"
One of the env variables that control the display eastereggs, it acts as a negative switch, ie: setting it to 1 will disable the display of eastereggs while leaving this var unset or setting it to 0 will allow the display of eastereggs.

.TP
.B "EASTEREGGS"
One of the env vars that control the display of eastereggs, it is a switch type variable, setting it to 0 will disable the display of eastereggs while setting it to 1 or leaving it unset will enable the display of eastereggs.

.TP
.B "SHITPOSTS"
One of the env variables that control the display eastereggs, it acts as a switch, ie: setting it to 1 will enable the display of the shitpost type eastereggs while leaving this var unset or setting it to 0 will disable the display of shitposts, this means unlike the regular eastereggs the shitposts are OPT-IN.

.TP
.B "JOBLESS_FAT_BASTARD_STOP_BEGGING_FOR_MONEY"
This is the env var to control the display of the donations message, it acts as a negative switch, setting it to 1 will disable the donations message completely, leaving it unset or setting as 0 will allow the donations message to be displayed.

.SH TEMPLATES
.P
Wal supports user defined template files. What this means is that similar to the export files in the
.B $XDG_CACHE_HOME/wal/
directory, you can create your own files which will be processed by pywal and spit out in this directory.
.P
Subdirectories with templates are also supported inside the templates dir.
.P
The user defined files are read from
.B $XDG_CONFIG_HOME/wal/templates/
and are exported to
.B $XDG_CACHE_HOME/wal/
under the same name. Using this feature you can also overwrite the default export files

.SH MISC
.P
The section for other important information to using pywal16 that does not fit a specific section.

.TP
.B "XDG_SESSION_TYPE, DISPLAY, WAYLAND_DISPLAY"
The
.B XDG_SESSION_TYPE
env var is needed to set the wallpaper when running an environment consisting of a standalone x11 window manager or wayland compositor (i.e. not a full desktop environment).
If 
.B XDG_SESSION_TYPE
is not available the
.B DISPLAY
and
.B WAYLAND_DISPLAY
vars are used as fallback to detect the active display protocol. This avoids relying only on
.B XDG_SESSION_TYPE
which may be set to "tty" on some setups (for example when starting X manually with startx) or not set at all. If
.B WAYLAND_DISPLAY
is set, the session is treated as Wayland. If
.B DISPLAY
is set (and
.B WAYLAND_DISPLAY
is not), the session is treated as X11.

This is necessary because some setup may have the packages for (example) BOTH i3 and sway, so in previous releases swaybg had priority over feh and would be ran as wallpaper setter regardless of the user currently running i3 or sway, these checks allow us to follow the corresponding chain of possible wallpaper setters.

.TP
.B "ImageMagick 7 Policy"
Back in about march 2024 ImageMagick 7 changed it's default policy to not allow writing to STDOUT, just locate the file 
.B "/etc/ImageMagick-7/policy.xml"
(if that is not the name in your distro consult the package file list) then find the line
.B '<policy domain="path" rights="none" pattern="-"/>'
and change rights="none" to rights="write"
For convenience you could use the following sed command:
sudo sed 's/policy domain="path" rights="none" pattern="-"/policy domain="path" rights="write" pattern="-"/' /etc/ImageMagick-7/policy.xml

Worth noting that in distros like debian and derivates the ImageMagick policy is modified at packaging time to allow writes to STDOUT.

.SH ORIGINAL AUTHOR
.EX
.EE
Dylan Araps            <dylan.araps at gmail.com>
.SH MAINTAINER
.EX
eylles                 <ed.ylles1997 at gmail.com>
.EE
.SH CONTRIBUTORS
.EX
For a list of contributors, run `git shortlog -s` in the nsxiv git repository.
.EE
.SH HOMEPAGE
.TP
Source code:
.EE
https://github.com/eylles/pywal16
.EX


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.8.14] - 2026-01-30
Fixes:
- Fix wal backend issues.
- Fix session type detection.

NEW:
- Add `awww` support.
- Add section to manpage about XDG_SESSION_TYPE and how the session is detected.

## [3.8.13] - 2025-12-13
Fixes:
- Fix issue where pywal16 would try to use wayland wallpaper setters in an x11 environment.

## [3.8.12] - 2025-11-28
Fixes:
- Fix regression from 3.8.11 where some images would crash the wal backend.

New:
- Added ENV VAR to disable the donations message, consult the manpage.

## [3.8.11] - 2025-10-20
Fixes:
- Allow the wal backend to work with just 8 colors instead of requiring all 16
- Template export parser

New:
- ENV vars to control the printing of the easteregg copypastas and texts
- Unit tests on push and pr
- dual method for cols16

## [3.8.10] - 2025-09-29
Fixes:
- Use exact name for waybar and polybar.
- More reliable imagemagick output parsing.
- Make the behaviour of the --saturation flag more intuitive.

New:
- `foot` consider alpha for the default template.
- `ultrakill` palette generation.

## [3.8.9] - 2025-05-31
Fixes:
- Fix the fix that was not fixed with `--out-dir`.

## [3.8.8] - 2025-05-31
Fixes:
- Error during color export.

## [3.8.7] - 2025-05-31
Fixes:
- Ensure alpha value is the correct type and is within valid range.
- Increase saturation of very dark theme background colors.

New:
- Prefer the alpha value from flag over the theme alpha.
- Allow the `--saturate` flag to affect colors 0 and 8.
- Added the `--out-dir` flag to allow changing export dir, it overrides the value of '$PYWAL_CACHE_DIR'.

## [3.8.6] - 2025-04-14
Fixes:
- `pywal` no longer crashes on windows when `colorama` is not present.

## [3.8.5] - 2025-03-31
New:
- Added tokyonight colorschemes.
- Set `alpha_num` value from command line even when loading theme files.
- Added mechanism to live reload nvim colors through `nvim-colo-reload`.

## [3.8.4] - 2025-03-25
New:
- Added rosepine colorschemes.
- Support reloading of mako.

## [3.8.3] - 2025-03-16
New:
- `retroarch` rgui colors template.
- `termux` support.

Fixes:
- `KDE` 6.3.3 error where pywal would crash upon failing to set the wallpaper.

## [3.8.2] - 2025-03-03
Fixes:
- Hyprland and Hyprpaper support actually this time.

    New:
- Fuzzel template.
- Tofi template.
- Mako template.
- Zathura template.
- webp image format support.

## [3.8.1] - 2025-02-23
Fixes:
- Hyprland and Hyprpaper support working properly.

New:
- Hyprland colors template.
- dwl colors template.
- Added support for `wbg`.
- Added catpuccin colorschemes.

## [3.8.0] - 2025-02-15

New features:
- Added Dynamic shade conversion, using the `--cols16` flag will convert 9 shades themes to 16 shades.
- Added Nested template directories, subdirectories with templates can be created inside the templates dir.
- Added support for pywalfox.
- Added support for waybar.
- Added support for hyprpaper.
- Added Dracula colorschemes.
- Added ghostty template.
- Added base16-kanagawa colorscheme.
- Added color template for dwm urgent window patch.

Bug fixes and Improvements:
- Better colors for Fish.
- Restrict colorama import only to ms windows.
- Change dwm selection background to match dmenu.

Removals:
- gtk reloading, python 2 is obsolete and that method worked only for gtk2.

## [3.7.2] - 2024-11-29

- make color.get and color.cache_fname API more backwards compatible.

## [3.7.1] - 2024-11-27

- Use test_file3.json in test_export.py.

## [3.7.0] - 2024-11-27

- Added `foot` terminal emulator template.
- Added `zed` template.
- Remove redundant template `colors.hs`.
- Add `--contrast` option.
- Enable MS Windows color output.
- Use kwargs for the cols16 and contrast internal API.
- Improved info log to include template directory paths.
- Add optional dependencies to setup.py.
- Made `checksum` available to the colorscheme dict.
- Added `adjust_alpha` function to be used in templates.
- Added support for MacOS Sonoma.
- Added manpage `wa.1`.
- Changed formatting to use `BLACK` formatter.

## [3.6.0] - 2024-07-04

- Fixed empty message on KDE plasma when sending sequences.
- Add the `okthief` backend.
- Fix issue with `fcntl` in unsupported platforms.
- Explicitly add the method by which 16 colors scheme is created to the file name.
- Fix the darkness of color 0
- Fix the darkness and saturation of colors 7 and 8

## [3.5.3] - 2024-06-02

- fixed issue with `wal` backend on arch where it would fail to generate a colorscheme.

## [3.5.3] - 2024-05-07

- Fixed issue with `wal` backend sometimes reading garbage data form imagemagick.
- Added notice for Imagemagick 7 policy change, fix on wiki.
- Added custom `foreground` argument just like custom `background` argument.
- Change open files to write in non-blocking mode.

## [3.5.2] - 2024-05-06

- modified generic_adjust and added a pre-adjust to the colorthief backend to match legacy background colors

## [3.5.1] - 2024-04-20

- Added wallpaper checksum generation and comparison for cached color schemes.
- Added toml template for new `alacritty` versions.

## [3.5.0] - 2024-02-xx

- Added rgb color template.
- Added support for `swww`.
- Added wallpaper override when using a theme.
- Change `alacritty` template to match v11.0+ config.
- Change wallpaper path on microsoft windows uses bytestring.
- Change when pywal is used as module allow using user templates.
- Added swaybg wallpaper support.
- Change `--cols16` flag now supports arguments `darken` and `lighten` to generate the color palette.

## [3.4.0] - 2022-10-31

- Fixed template exporting issue, rewriting may be necessary for a proper fix.
- Added new properties red_dec, green_dec and blue_dec.
- Added new properties red_hex, green_hex and blue_hex.

pywal/backends/wal.py
"""
Generate a colorscheme using imagemagick.
"""

import logging
import re
import shutil
import subprocess
import sys

from .. import colors


def imagemagick(color_count, img, magick_command):
    """Call Imagemagick to generate a scheme."""
    flags = [
        "-resize",
        "25%",
        "-colors",
        str(color_count),
        "-unique-colors",
        "txt:-",
    ]
    img += "[0]"

    try:
        output = subprocess.check_output(
            [*magick_command.split(), img, *flags], stderr=subprocess.STDOUT
        ).splitlines()
    except subprocess.CalledProcessError as Err:
        logging.error("Imagemagick error: %s", Err)
        logging.error(
            "IM 7 disables stdout by default, check the manual or wiki to fix."
        )
        sys.exit(1)
    return output


def has_im():
    """Check to see if the user has im installed."""
    magick_commands = []

    if shutil.which("magick"):
        magick_commands.extend(["magick", "magick convert"])

    if shutil.which("convert"):
        magick_commands.append("convert")

    if len(magick_commands) > 0:
        return magick_commands

    logging.error("Imagemagick wasn't found on your system.")
    logging.error("Try another backend. (wal --backend)")
    sys.exit(1)


def gen_colors_with_command(
        img, magick_command, beginning_color_count=16, iteration_count=20
        ):
    """Iteratively attempt to generate a 16-color palette
    using a specific Imagemagick command."""
    hex_pattern = re.compile(r"#[A-F0-9]{6}", re.IGNORECASE)

    max_color_count = beginning_color_count + iteration_count - 1
    for color_count in range(
            beginning_color_count, beginning_color_count + iteration_count
            ):
        raw_output = imagemagick(color_count, img, magick_command)
        hex_colors = [
            hex_pattern.search(str(col)).group()
            for col in raw_output
            if hex_pattern.search(str(col))
        ]

        if len(hex_colors) >= 16:
            break

        if color_count < max_color_count:
            logging.warning(
                    "Imagemagick couldn't generate a "
                    f"palette with {magick_command}."
                    )
            logging.warning(
                    f"Trying a larger palette size {color_count}."
                    )
        else:
            logging.error(
                    "Imagemagick couldn't generate a suitable palette "
                    f"with {magick_command}."
                    )
            logging.warning(
                    "Will try to do palette concatenation, "
                    "good results not guaranteed!"
                    )
            while len(hex_colors) < 16:
                hex_colors.extend(hex_colors)
    return hex_colors


def gen_colors(img):
    """Try each Imagemagick command until a color palette
    is successfully generated."""
    magick_commands = has_im()

    for magick_command in magick_commands:
        logging.debug(f"Trying {magick_command}...")

        hex_colors = gen_colors_with_command(img, magick_command)

        if not hex_colors:
            logging.warning(
                    f"Failed to generate colors with {magick_command}."
                    )
            continue

        return hex_colors

    raise RuntimeError(
            "Failed to generate color palette from "
            f"{img} with these commands: {magick_commands}"
            )


def adjust(cols, light, **kwargs):
    """Adjust the generated colors and store them in a dict that
    we will later save in json format.
    :keyword-args:
    -    c16: use 16 colors through specified method - [ "lighten" | "darken" ]
    """
    if "c16" in kwargs:
        cols16 = kwargs["c16"]
    else:
        cols16 = False
    raw_colors = cols[:1] + cols[8:16] + cols[8:-1]

    return colors.generic_adjust(raw_colors, light, c16=cols16)


def get(img, light=False, **kwargs):
    """Get colorscheme.
    :keyword-args:
    -    c16: use 16 colors through specified method - [ "lighten" | "darken" ]
    """
    if "c16" in kwargs:
        cols16 = kwargs["c16"]
    else:
        cols16 = False
    colors = gen_colors(img)
    # it is possible we could have picked garbage data
    garbage = "# Image"
    if garbage in colors:
        colors.remove(garbage)
    return adjust(colors, light, c16=cols16)
